### PR TITLE
Add trilium server and module

### PR DESCRIPTION
**Pull request by kampka**
###### Motivation for this change

This change adds the trilium server package and a module to run it.
It also refactors the trilium package as such so its easier to keep the metadata and version of both packages in sync.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @dtzWill @emmanuelrosa

### CODE REVIEW

File: nixos/modules/services/web-apps/trilium.nix
Comment (aanderse) [inline nixos/modules/services/web-apps/trilium.nix]: Does this option provide value to you? Do you have a use case for this option where you wouldn't use the default you might share?

File: nixos/modules/services/web-apps/trilium.nix
Comment (aanderse) [inline nixos/modules/services/web-apps/trilium.nix]: Would you consider adding a `settings` option as described by https://github.com/NixOS/rfcs/pull/42?

File: nixos/modules/services/web-apps/trilium.nix
Comment (aanderse) [inline nixos/modules/services/web-apps/trilium.nix]: `Environment.TRILLIUM_DATA_DIR = cfg.dataDir;`

**Review by aanderse — CHANGES_REQUESTED** [review]
I have only reviewed the nixos module. Further to the points mentioned below there are inconsistencies between your spelling `trillium` vs `trilium`.

File: nixos/modules/services/web-apps/trilium.nix
Comment (aanderse) [inline nixos/modules/services/web-apps/trilium.nix]: I'm starting to wonder if `DynamicUser` would be a good fit here. I think it might, if you were to hard code `dataDir`.

File: nixos/modules/services/web-apps/trilium.nix
Comment (kampka) [reply nixos/modules/services/web-apps/trilium.nix]: Done

File: nixos/modules/services/web-apps/trilium.nix
Comment (kampka) [reply nixos/modules/services/web-apps/trilium.nix]: We have already discussed this once on this [PR](https://github.com/NixOS/nixpkgs/pull/60788#issuecomment-491436203).

Bottom line: I like my system to not break because some default changes without me noticing.
If the Postgres socket directory had been a configuration option instead of a magic value, changing it from `/tmp` to `/run/postgres` would not have required a backwards incompatible change with 19.09 because one could just reference it as `${services.porstgres.socketDir}` in depending services.
With data, I think it's vital that this is controllable because backups will just silently continue to run on the old paths without braking, for example. Same is true for mount points, quotas, (sub-) volumes, etc.

**Review by aanderse — CHANGES_REQUESTED** [review]

File: nixos/modules/services/web-apps/trilium.nix
Comment (aanderse) [reply nixos/modules/services/web-apps/trilium.nix]: Right. Thanks for reminding me. We continue to respectfully disagree on this topic, but as I'll continue to mention I'm happy to accept modules which include a `dataDir` option.

File: nixos/modules/services/web-apps/trilium.nix
Comment (kampka) [reply nixos/modules/services/web-apps/trilium.nix]: Looking at the RFC, there seems to be a lot of boilerplate still missing.
I am reading your comment as "Add a nested settings sub-option that can later be changed to use types.settings.ini / genINI without breaking compatibility." Did I get that right?

File: nixos/modules/services/web-apps/trilium.nix
Comment (aanderse) [reply nixos/modules/services/web-apps/trilium.nix]: I have also suggested databases have a `socket` option, but `readOnly`.

File: nixos/modules/services/web-apps/trilium.nix
Comment (aanderse) [reply nixos/modules/services/web-apps/trilium.nix]: Given our discussion above we can drop this point.

File: nixos/modules/services/web-apps/trilium.nix
Comment (kampka) [reply nixos/modules/services/web-apps/trilium.nix]: > I have also suggested databases have a `socket` option, but `readOnly`.

I would be very much in favor of this!

**Comment by kampka** [comment]
@aanderse Thanks very much for the quick review, I appreciate it :)

File: nixos/modules/services/web-apps/trilium.nix
Comment (aanderse) [reply nixos/modules/services/web-apps/trilium.nix]: That is the essence of my suggestion, yes.

I thought it would be nice to add some flexibility to the module. I didn't go through the full list of available options in upstream documentation, but I saw some options for serving over https which I assume would be of value to some people.

I've never used this application, or configured it, so if there really isn't anything of value that is a valid answer too.

Comment (kampka) [reply nixos/modules/services/web-apps/trilium.nix]: All remaining options concern https which I left out because it is so easy to just configure SSL offloading via nginx/caddy/etc. on NixOS that I didn't think anyone would bother, but I'm happy to add those if requested.

As for the flexibility argument, I agree that RFC 42 will give us that and I am absolutely in favor of adding it's implications to this module. I think however that a lot is still missing and compatibility depends highly on how the actual structure and semantics of the resulting types, especially how hierarchies are going to be represented. Since the configuration file in this case is INI format, there are probably close to a million opinions as to how that could go.

I'd propose filing a bug for this blocked RFC 42 that I'll happily tackle once that's formalized.

Comment (aanderse) [reply nixos/modules/services/web-apps/trilium.nix]: If the only other options are related to web then I agree a reverse proxy seems like a good idea. Any interest in adding that option to this PR? Can always add at a later date if not.

:+1:

File: nixos/modules/services/web-apps/trilium.nix
Comment (aanderse) [inline nixos/modules/services/web-apps/trilium.nix]: You should use `systemd.tmpfiles.rules` instead of a separate service.

**Review by aanderse — CHANGES_REQUESTED** [review]
For some reason, presumably user error :confused:, my mobile browser seems to have lost one of the comments I made on this PR. So here we go again... sorry for the noise.

File: nixos/modules/services/web-apps/trilium.nix
Comment (kampka) [reply nixos/modules/services/web-apps/trilium.nix]: Right, forgot about that RFC completely.

Done.

File: nixos/modules/services/web-apps/trilium.nix
Comment (kampka) [reply nixos/modules/services/web-apps/trilium.nix]: Added a configuration option for nginx, I have no experience with other servers I'm afraid.

**Comment by kampka** [comment]
> For some reason, presumably user error confused, my mobile browser seems to have lost one of the comments I made on this PR. So here we go again... sorry for the noise.

@aanderse for some reason GH has recorded this message as a change request. 
Is there anything we need to do there to not have it block the merge?

File: nixos/modules/services/web-apps/trilium.nix
Comment (aanderse) [inline nixos/modules/services/web-apps/trilium.nix]: Having never used `nginx` or the module: can you use `locations<name>.proxyPass` here in addition to `extraConfig`? Don't consider this a change request, but informational for me because I'm working on a PR were I need to better understand the `nginx` module.

File: nixos/modules/services/web-apps/trilium.nix
Comment (aanderse) [inline nixos/modules/services/web-apps/trilium.nix]: Why `exec`? `systemd.services.<name>.Type`?

**Review by aanderse — CHANGES_REQUESTED** [review]
Mostly fairly minor things at this point.

File: nixos/modules/services/web-apps/trilium.nix
Comment (kampka) [reply nixos/modules/services/web-apps/trilium.nix]: I find the spacing improves readability as it aligns the different fields.
I can live without it though if required.

File: nixos/modules/services/web-apps/trilium.nix
Comment (kampka) [reply nixos/modules/services/web-apps/trilium.nix]: Moved to `ExecStart`, seems simplest.

File: nixos/modules/services/web-apps/trilium.nix
Comment (kampka) [reply nixos/modules/services/web-apps/trilium.nix]: Removed the default from hostName, making this obsolete

File: nixos/modules/services/web-apps/trilium.nix
Comment (kampka) [reply nixos/modules/services/web-apps/trilium.nix]: Done

File: nixos/modules/services/web-apps/trilium.nix
Comment (kampka) [reply nixos/modules/services/web-apps/trilium.nix]: Done

File: nixos/modules/services/web-apps/trilium.nix
Comment (kampka) [reply nixos/modules/services/web-apps/trilium.nix]: Done

File: nixos/modules/services/web-apps/trilium.nix
Comment (kampka) [reply nixos/modules/services/web-apps/trilium.nix]: Your example had a capital `E`, making it not compile. Thanks for pointing out the obvious (RTFM).

File: nixos/modules/services/web-apps/trilium.nix
Comment (kampka) [reply nixos/modules/services/web-apps/trilium.nix]: You can use both `proxyPass` and `extraConfig`, I didn't feel like there is a huge need to do this here as the extraConfig block is already quite huge by itself. Refactored it anyways.

Comment (aanderse) [reply nixos/modules/services/web-apps/trilium.nix]: Thanks for explaining.

File: nixos/modules/services/web-apps/trilium.nix
Comment (aanderse) [reply nixos/modules/services/web-apps/trilium.nix]: I find my brain often jumbles `systemd` names with NixOS option names. Sorry for the confusion.

**Review by aanderse — APPROVED** [review]
Module looks good to me! Thanks.

I haven't reviewed anything but the module so I'll wait for review from @dtzWill, @emmanuelrosa, or anyone else who cares to provide it. If no one jumps on this in the next couple days ping me and I'll take a look at the package changes.

**Comment by kampka** [comment]
@aanderse Thanks a lot for this comprehensive and instructive (also: timely ;) ) review !

**Comment by emmanuelrosa** [comment]
I tested trilium-desktop and trilium-server. Both worked :)

I used a NixOS container to install trilium-server and then connected from the host:

**trilium-container.nix**
```
# Edit this configuration file to define what should be installed on
# your system.  Help is available in the configuration.nix(5) man page
# and in the NixOS manual (accessible by running ‘nixos-help’).

{ config, pkgs, lib, ... }:

{

   services.trilium-server = {
    enable = true;
    nginx = {
      enable = true;
      hostName = "trilium.example.com";
    };
  };

  networking.firewall.enable = false;
}
```

`nixos-container create trilium --config-file trilium-container.nix --host-address 192.168.1.108 --local-address 10.233.0.100`

**Comment by emmanuelrosa** [comment]
I forgot to mention that trilium-server is not logging to journald. Instead it's logging to a file in `dataDir`.

**Comment by kampka** [comment]
> I forgot to mention that trilium-server is not logging to journald. Instead it's logging to a file in `dataDir`.

It is actually doing both, although not quite consistently (https://github.com/zadam/trilium/blob/c4d5060a0b9f2066e661725c6d32ed2f016b8b71/src/services/log.js).
It's also not just the server, the desktop version does this too in `~/.local/share/trilium-data/log`.
As there is no way to configure this, I see only the option to patch it on our side.
If anyone thinks it's worth it I'll prepare a patch for review

**Comment by kampka** [comment]
@emmanuelrosa I've added a patch for the server version that removes the file logger and logs to console only now.
The desktop version is a pre-build electron app, I don't see a good way to patch that without building from source.

File: pkgs/applications/office/trilium/default.nix
Comment (emmanuelrosa) [inline pkgs/applications/office/trilium/default.nix]: You can use the patchPhase built into stdenv to apply the patch. Note, the patch will apply before the installPhase:

```
diff --git a/pkgs/applications/office/trilium/default.nix b/pkgs/applications/office/trilium/default.nix
index 579fe1a7c58..2c8c2b47eab 100644
--- a/pkgs/applications/office/trilium/default.nix
+++ b/pkgs/applications/office/trilium/default.nix
@@ -91,6 +91,8 @@ in {
       libxkbfile
     ];
 
+    patches = [ ./0001-Use-console-logger-instead-of-rolling-files.patch ];
+
     installPhase = ''
       mkdir -p $out/bin
       mkdir -p $out/share/trilium-server
@@ -104,7 +106,6 @@ in {
       chmod a+x $out/bin/trilium-server
 
       cd $out/share/trilium-server
-      patch -p1 < "${ ./0001-Use-console-logger-instead-of-rolling-files.patch }"
     '';
   };
 }
```

File: pkgs/applications/office/trilium/default.nix
Comment (emmanuelrosa) [inline pkgs/applications/office/trilium/default.nix]: I keep thinking there must be an idiomatic way of creating the launch script so that it's not necessary to generate a script within a script.

I think the problem is that upstream doesn't provide a script to run trilium-server. Their recommended deployment method seems to be Docker, in which case they run node, just like in this script. That leads me to think a patch should be used to add a launch script, but then the script would need to be patched up to take the Nix store paths into account.

Another way is to use `pkgs.writeScript`, which would create the script in the Nix store, but not at a path of your choosing. Which means you'd essentially have three packages: trilium-desktop, trilium-server, and (for example) trilium-server-launcher. This might be overkill ;)

A simple change would be to move this into the postFixup phase. It would have the same effect, but this is the phase where typically things like wrapper scripts are created:

```
diff --git a/pkgs/applications/office/trilium/default.nix b/pkgs/applications/office/trilium/default.nix
index 2c8c2b47eab..28956d2a7c7 100644
--- a/pkgs/applications/office/trilium/default.nix
+++ b/pkgs/applications/office/trilium/default.nix
@@ -98,14 +98,15 @@ in {
       mkdir -p $out/share/trilium-server
 
       cp -r ./* $out/share/trilium-server
+    '';
+
+    postFixup = ''
       cat > $out/bin/trilium-server <<EOF
       #!${stdenv.cc.shell}
       cd $out/share/trilium-server
       exec ./node/bin/node src/www
       EOF
       chmod a+x $out/bin/trilium-server
-
-      cd $out/share/trilium-server
     '';
   };
 }
```

File: pkgs/applications/office/trilium/default.nix
Comment (kampka) [reply pkgs/applications/office/trilium/default.nix]: Done

File: pkgs/applications/office/trilium/default.nix
Comment (kampka) [reply pkgs/applications/office/trilium/default.nix]: Moved to `postFixup`. 
I thought about using `pkgs.writeScript` but in the end the result would be more complex I believe. In the end, the service does not need either, it could just as well run `node` directly so this is mostly about user convenience. Providing an extra derivation that generated the executable script seems counter-intuitive for that.

Comment (emmanuelrosa) [reply pkgs/applications/office/trilium/default.nix]: Yes, that's true.

To sum it up, I like how you put this all together. It's all very clean, precise, and functional. And it's got tests!!!

Thank you, very much.

Comment (kampka) [reply pkgs/applications/office/trilium/default.nix]: > Thank you, very much.

Thank you for reviewing :+1:

File: nixos/modules/services/web-apps/trilium.nix
Comment (aanderse) [reply nixos/modules/services/web-apps/trilium.nix]: @kampka maybe you accidentally pushed something wrong because the `default` value and assertion still show.

Comment (kampka) [reply nixos/modules/services/web-apps/trilium.nix]: @aanderse Yeah, something seems to have gone wrong there, should be fixed now.
Good catch, thanks :+1:

**Comment by kampka** [comment]
@GrahamcOfBorg build trilium-server
@GrahamcOfBorg build trilium-desktop
@GrahamcOfBorg test trilium-server

**Comment by kampka** [comment]
@GrahamcOfBorg build trilium-server
@GrahamcOfBorg build trilium-desktop
@GrahamcOfBorg test trilium-server

**Comment by kampka** [comment]
Rebased against master and updated the package version once more.
Anyone willing to push the button? :)

**Comment by aanderse** [comment]
@kampka oops, sorry for the delay. I didn't notice the comments from @emmanuelrosa.

Speaking of which... @emmanuelrosa do you mind hitting the approve button please? I took a very quick look over the package but I would like you as the listed maintainer to give it the official thumbs up.

@kampka can you please address the `aarch64` failure?

**Comment by kampka** [comment]
@GrahamcOfBorg build trilium-desktop
@GrahamcOfBorg build trilium-server
@GrahamcOfBorg test trilium-server

**Comment by kampka** [comment]
@GrahamcOfBorg build trilium-desktop
@GrahamcOfBorg build trilium-server
@GrahamcOfBorg test trilium-server

**Comment by kampka** [comment]
@GrahamcOfBorg test trilium-server

**Comment by kampka** [comment]
@aanderse sorry for the noise, the fact that ofBorg does an "empty" run on unsupported archs threw me of there.
Pinned the executables and tests to x86_64-linux

**Comment by emmanuelrosa** [comment]
I don't have an approval button, but nevertheless, I approve :)

**Comment by aanderse** [comment]
> Pinned the executables and tests to x86_64-linux

I thought if you marked a package as `x86_64-linux` tests would only run for that platform. I don't think you need to specifically mention that in the list of tests... but I'm not sure off the top of my head.

> I don't have an approval button, but nevertheless, I approve :)

If you review a package you should have the ability to approve, comment, or request changes. Thank you for explicitly stating, though :smile: 

As an aside... you are marked as a maintainer for this package yet you aren't a member of the NixOS github project :thinking: If this is not by choice we should ping the maintainer of that code because it was my understanding that everyone listed as a maintainer was supposed to be added to the project at one point. Let me know if you want this corrected.

**Review by aanderse — APPROVED** [review]
:+1:

Thanks @kampka! ...and sorry for the long drawn out process.

**Comment by emmanuelrosa** [comment]
@aanderse, I was involved in developing the original package. I suppose that's why I'm listed as a maintainer. I'd be honored to be a member of the NixOS project.

**Comment by kampka** [comment]
@aanderse @emmanuelrosa thanks for taking the time to review this guys :+1:

**Comment by aanderse** [comment]
@kampka no problem. Thanks for the contribution :tada:

**Pull request merged**
